### PR TITLE
[Site Editor]: Fix multiple templates creation while in the process of creating one

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Mentioned by @annezazu here: https://github.com/WordPress/gutenberg/issues/43877#issuecomment-1246070721

Currently when we're in the process of creating a template in Site Editor, a user can trigger the creation flow for multiple templates. This PR fixes that by adding an overlay with a spinner during the request to create a template.

Based on the below  @jameskoster 's https://github.com/WordPress/gutenberg/pull/44146#issuecomment-1246437841

## Testing Instructions
1. In site editor you should be able to create a single template at the time, no matter how many times you click a menu item or button to create one.
2. The issues is more visible on slow networks.

### Before

https://user-images.githubusercontent.com/26996883/190030104-dacafbf2-3d90-46fe-9120-0f5ff2ddbc8a.mov

### After
I'm not creating the actual templates in the below video to showcase the overlay.

https://user-images.githubusercontent.com/16275880/190187590-52e3f3e7-c5f8-45f0-9fb0-e8b8f06c1b64.mov

### CODE REVIEW

**Comment by jameskoster** [comment]
Nice work. 

I don't think the Snackbar is the correct approach though, for a couple of reasons:

1. On quicker connections it's strange to see it considering the creation process is almost instant
2. It doesn't prohibit multiple clicks in the templates menu (or modal), so the experience still feels a bit awkward in that case

As an alternative could we disable the menu / modal, with an overlay and spinner?

new-template

category

If it's easier, a full-screen overlay / spinner applied in both cases could work too.

**Comment by jameskoster** [comment]
> If it's easier, a full-screen overlay / spinner applied in both cases could work too.

template

**Comment by ntsekouras** [comment]
@jameskoster I added overlays where needed. If we're in a Modal flow the overlay is also shown in the DropDown as a user could click outside during creation and click other menu items.

File: packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal.js [hunk 1]
```diff
@@ -16,27 +16,38 @@ import {
...
+				{
+					slug:
+						'wp-custom-template-' +
+						kebabCase( title || defaultTitle ),
+					title: title || defaultTitle,
+				},
+				false
+			);
+		} finally {
+			setIsBusy( false );
```
Comment (ntsekouras) [inline packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal.js:49]: This is an unrelated small bug I discovered while testing and `throwing` error.

File: packages/edit-site/src/components/add-new-template/new-template.js [hunk 2]
```diff
@@ -176,66 +184,73 @@ export default function NewTemplate( { postType } ) {
...
-								onClick={ () =>
-									setShowCustomGenericTemplateModal( true )
-								}
-							>
-								{ __( 'Custom template' ) }
-							</MenuItem>
-						</MenuGroup>
-					</NavigableMenu>
+					<>
+						{ isCreatingTemplate && <CreatingTemplateOverlay /> }
```
Comment (ntsekouras) [inline packages/edit-site/src/components/add-new-template/new-template.js]: This is the only addition here in the `DropdownMenu`. The rest is whitespace.

**Comment by jameskoster** [comment]
Thanks Nik, I think that's better. There is one issue that I didn't anticipate though (sorry)... 

On a very slow connection, after clicking to create a new template, you can close the modal + click to edit a different template, all before the template has been created.

https://user-images.githubusercontent.com/846565/190194902-4818a05d-233c-41f8-8ce0-17c1c0dfa202.mp4

Perhaps a [full-screen overlay](https://github.com/WordPress/gutenberg/pull/44146#issuecomment-1246442992) would be better after all?

I'm also wondering if the work being done in https://github.com/WordPress/gutenberg/pull/42525 might overlap here?

---

Edit: Do we need to do something similar with deletion as well? On a slow connection you can repeatedly delete the same template before the deletion actually takes place 🙈

**Comment by annezazu** [comment]
Thank you so much for splitting this out and getting started on a fix 🙌🏼 +1 to needing the same for deletion.

**Comment by ntsekouras** [comment]
>I'm also wondering if the work being done in https://github.com/WordPress/gutenberg/pull/42525 might overlap here?

I don't think so, since that PR handles the loading state before rendering blocks. --cc @tyxla 
Maybe the overlay could be consistent eventually, but I'd leave that decision to designers.

I just pushed some changes to try the fullscreen overlay for the template creation for now.

**Comment by jameskoster** [comment]
Yeah this is feeling pretty good now. I'd say it's in a spot to land and refine. 

A tiny nitpick... would it be possible to have the overlay fill the entire viewport? The small sliver of UI peeking through at the top is a bit awkward.

Screenshot 2022-09-15 at 15 00 41

**Review by jameskoster — APPROVED** [review]
Seems good to me, thanks!
